[CK_TILE] Add GetName for GEMM kernels

aledudek opened the pull request:
## Proposed changes

Add GetName for GEMM kernels to easily print its name and a few other details.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

Small change. I am also open to print other kernel parameters.

aledudek committed:
[CK_TILE] Add GetName functions for Gemm Kernels

aledudek committed:
[CK_TILE] Add GetName for grouped gemm

afagaj commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -75,6 +75,40 @@ struct GemmKernel
+
+    CK_TILE_HOST static std::string GetName()
+    {
+#define _SS_ std::string
+#define _TS_ std::to_string
+        // clang-format off
+        using P_ = GemmPipeline;
+
+            auto prec_str = [&] () {
+            std::string base_str = _SS_(t2s<ADataType>::name);
+            if (!std::is_same_v<ADataType, BDataType>) {
+                base_str += _SS_("_") + _SS_(t2s<BDataType>::name);
+            }
+            return base_str;
+        }();
+
+        return _SS_("gemm_") + _SS_(prec_str) + "_" +
+                _TS_(P_::kMPerBlock) + "x" + _TS_(P_::kNPerBlock) + "x" + _TS_(P_::kKPerBlock) + "_" +
+                _TS_(P_::VectorSizeA) + "x" + _TS_(P_::VectorSizeB) + "x" + _TS_(P_::VectorSizeC) + "_" +
+                _TS_(P_::kPadM) + "x" + _TS_(P_::kPadN) + "x" + _TS_(P_::kPadK);
+#undef _SS_
+#undef _TS_
```
Can you achieve the same result without using the pre-processor? (e.g. `using _SS_ = std::string` instead of `#define _SS_ std::string`)

aledudek replied:
Yes, we can replace `#define _SS_ std::string` with `using`, but we can not replace `#define _TS_ std::to_string` with an `using` statement. Changed the _SS_ one

aosewski commented on include/ck_tile/ops/gemm/kernel/gemm_kernel.hpp:
```diff
@@ -75,6 +75,40 @@ struct GemmKernel
     static constexpr auto I1 = number<1>();
     static constexpr auto I2 = number<2>();
 
+    // clang-format off
+    template <typename T> struct t2s;
+    template <> struct t2s<float> { static constexpr const char * name = "fp32"; };
+    template <> struct t2s<fp16_t> { static constexpr const char * name = "fp16"; };
+    template <> struct t2s<bf16_t> { static constexpr const char * name = "bf16"; };
+    template <> struct t2s<fp8_t> { static constexpr const char * name = "fp8"; };
+    template <> struct t2s<bf8_t> { static constexpr const char * name = "bf8"; };
+    template <> struct t2s<int8_t> { static constexpr const char * name = "int8"; };
+    // clang-format on
```
This is a utility which shows up here and there. Lets put it in some utility header and reuse.

aosewski commented on include/ck_tile/ops/gemm/kernel/grouped_gemm_kernel.hpp:
```diff
@@ -44,6 +44,40 @@ struct GroupedGemmKernel
     using BDataType = remove_cvref_t<typename GemmPipeline::BDataType>;
     using CDataType = remove_cvref_t<typename EpiloguePipeline::ODataType>;
 
+    // clang-format off
+    template <typename T> struct t2s;
+    template <> struct t2s<float> { static constexpr const char * name = "fp32"; };
+    template <> struct t2s<fp16_t> { static constexpr const char * name = "fp16"; };
+    template <> struct t2s<bf16_t> { static constexpr const char * name = "bf16"; };
+    template <> struct t2s<fp8_t> { static constexpr const char * name = "fp8"; };
+    template <> struct t2s<bf8_t> { static constexpr const char * name = "bf8"; };
+    template <> struct t2s<int8_t> { static constexpr const char * name = "int8"; };
+    // clang-format on
```
As above this should be in common place.

aosewski reviewed (dismissed):
I think it would be beneficial to make it much more comprehensive. Lets add also the size of warp tiles next to block tile. Additionally I'd add similar functions to Pipelines and block gemms, so that later on during benchmarking we could easily get report with what instance we're working on.

aledudek committed:
[CK_TILE] Add GetName for gemm - review changes

aledudek committed:
Merge branch 'develop' into gemm_getname

aledudek committed:
Merge branch 'develop' into getname_gemm

aledudek committed:
Merge branch 'develop' into getname_gemm

aledudek committed:
[CK_TILE] Print also gemm problem pipeline and shape

aosewski commented on include/ck_tile/ops/common/utils.hpp:
```diff
@@ -0,0 +1,20 @@
+#include <string>
+
+#include "ck_tile/core.hpp"
+namespace ck_tile {
```
Add space above.

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_comp_v3.hpp:
```diff
@@ -77,6 +77,20 @@ struct GemmPipelineAgBgCrCompV3 : public BaseGemmPipelineAgBgCrCompV3<Problem>
+        using _SS_ = std::string;
+
+        return _SS_("pipeline_AgBgCrCompV3_") +
+                _TS_(MPerBlock) + "x" + _TS_(NPerBlock) + "x" + _TS_(KPerBlock) + "x" + _TS_(BlockSize) + "_" +
```
There's no need to duplicate `MPerBlock` , NPerBlock KPerBLock info here. It's in Tile Shape

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_problem.hpp:
```diff
@@ -35,6 +35,19 @@ struct GemmPipelineProblemBase
+        using _SS_ = std::string;
+
+        return _SS_("gemm_problem_") +
+                _TS_(VectorLoadSize) + "x" + _TS_(kBlockSize) + "_" +
```
We would need here scheduler info also.

aledudek committed:
[CK_TILE] Print also GemmPipelineScheduler

aledudek committed:
Merge branch 'develop' into getname_gemm

aosewski commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_scheduler.hpp:
```diff
@@ -71,3 +72,10 @@ inline std::ostream& operator<<(std::ostream& os, const ck_tile::TailNumber& s)
+
+inline std::string GemmPipelineSchedulerToString(const ck_tile::GemmPipelineScheduler& s)
+{
+    std::ostringstream oss;
+    oss << s;
+    return oss.str();
```
Wouldn't this print just integer value?

aosewski reviewed (commented):
Minor thing left.

mozga-amd commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_scheduler.hpp:
```diff
@@ -71,3 +72,10 @@ inline std::ostream& operator<<(std::ostream& os, const ck_tile::TailNumber& s)
+
+inline std::string GemmPipelineSchedulerToString(const ck_tile::GemmPipelineScheduler& s)
+{
+    std::ostringstream oss;
```
`thread_local` std::ostringstream oss ? then an object of this storage class is created when the thread of execution passes through its definition and is destroyed immediately when execution leaves the lexical.

aledudek replied:
Removed the function but added thread_local for the std::ostringstream oss; used in GetName()

mozga-amd commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_scheduler.hpp:
```diff
@@ -71,3 +72,10 @@ inline std::ostream& operator<<(std::ostream& os, const ck_tile::TailNumber& s)
     return os;
 }
+
+inline std::string GemmPipelineSchedulerToString(const ck_tile::GemmPipelineScheduler& s)
```
[[nodiscard]] fn?

aledudek replied:
Removing this function due to other changes

mozga-amd commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -57,6 +57,30 @@ struct BatchedGemmKernel : public GemmKernel<TilePartitioner_, GemmPipeline_, Ep
+        // clang-format off
+        using P_ = GemmPipeline;
+
+            auto prec_str = [&] () {
```
align? - here and others prec_str fn

mozga-amd commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_ag_bg_cr_scheduler.hpp:
```diff
@@ -71,3 +72,10 @@ inline std::ostream& operator<<(std::ostream& os, const ck_tile::TailNumber& s)
+
+inline std::string GemmPipelineSchedulerToString(const ck_tile::GemmPipelineScheduler& s)
+{
+    std::ostringstream oss;
```
oss.str("")? - it could be fine to clean buffer before

aledudek replied:
No need to clear the buffer here as it was just created. It would be needed when reusing oss another time

mozga-amd commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_problem.hpp:
```diff
@@ -35,6 +35,20 @@ struct GemmPipelineProblemBase
+
+        return _SS_("gemm_problem_") +
+                _TS_(VectorLoadSize) + "x" + _TS_(kBlockSize) + "_" +
+                _TS_(kPadM) + "x" + _TS_(kPadN) + "x" + _TS_(kPadK) + "_" +
```
What do you think about the concat() function, to pass Scheduler and other params?

aledudek replied:
What concat() function do you mean? I prefer using the + operator for such concatenation over using multiple strcat() or append() in this case

mozga-amd replied:
[[Consider]] Approach: A function that takes all parameters, performs concatenation and appropriate type casting internally, and then returns the resulting concatenated string... you can still use the + operator. By making it a variadic function, you will eliminate multiple + operators and multiple type castings, one `prec_str` fn for all kernels, single `::operator<<` using, single mapping-`#define` etc.
Proposition: 
```cpp
concat("pipeline_AGmemBGmemCRegV1_", "kMPerBlock", ....);
```
and concat()
```cpp
template <typename... Ts>
[[nodiscard]] auto concat(const Ts&... xs)
    -> std::enable_if_t<..... >, std::string>
```

mozga-amd commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -57,6 +57,30 @@ struct BatchedGemmKernel : public GemmKernel<TilePartitioner_, GemmPipeline_, Ep
     using BLayout          = typename Base::BLayout;
     using CLayout          = typename Base::CLayout;
 
+    CK_TILE_HOST static std::string GetName()
```
return const and [[nodiscard]]?

aledudek committed:
[CK_TILE] GetName - fixed Scheduler <<operator visibility

aledudek committed:
Merge branch 'develop' into getname_gemm

mozga-amd commented on example/ck_tile/03_gemm/gemm_basic.cpp:
```diff
@@ -87,7 +87,10 @@ float gemm_calc(const ck_tile::GemmHostArgs& args, const ck_tile::stream_config&
     if(s.log_level_ > 0)
     {
-        std::cout << "Launching kernel with args:"
+        std::cout << "Launching kernel: " << Kernel::GetName()
```
[[Consider]] What do you think about smoothing out the logs a bit and formatting them into blocks, to improve readability? Let's say:
```shell
Launching kernel with args:
Launching kernel:  Kernel::GetName()
shape:             CodegenGemmShape::GetName()
.....
```

mozga-amd commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_v1.hpp:
```diff
@@ -39,6 +39,34 @@ struct GemmPipelineAGmemBGmemCRegV1
     static constexpr bool kPadN = Problem::kPadN;
     static constexpr bool kPadK = Problem::kPadK;
 
+<<<<<<< HEAD
```
Some leftover from the merge.

mozga-amd commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_v1.hpp:
```diff
@@ -39,6 +39,34 @@ struct GemmPipelineAGmemBGmemCRegV1
+        return integer_divide_ceil(
+                   sizeof(ADataType) *
+                       Policy::template MakeALdsBlockDescriptor<Problem>().get_element_space_size(),
+                   16) *
```
Maybe it would be fine to make `16` as a variable?

mozga-amd commented on include/ck_tile/ops/common/utils.hpp:
```diff
@@ -0,0 +1,22 @@
+
+// clang-format off
+template <typename T> struct t2s;
+template <> struct t2s<float> { static constexpr const char * name = "fp32"; };
```
Maybe a more elegant name than `t2s` -> 'typeToStr`?

aledudek committed:
[CK_TILE] GetName info adjustments

aledudek committed:
Merge branch 'develop' into getname_gemm

aledudek committed:
[CK_TILE] GetName post-merge fix

aledudek committed:
[CK_TILE] GetName - add general concat function

mozga-amd commented on include/ck_tile/ops/gemm/kernel/grouped_gemm_kernel.hpp:
```diff
@@ -64,6 +64,25 @@ struct GroupedGemmKernel : public GemmKernel<TilePartitioner_, GemmPipeline_, Ep
+            return base_str;
+        }();
+
+        return concat("gemm_grouped_", prec_str, "_", P_::kMPerBlock, "x", P_::kNPerBlock, "x", P_::kKPerBlock, "_", P_::VectorSizeA, "x", P_::VectorSizeB, "x", P_::VectorSizeC, "_", P_::kPadM, "x", P_::kPadN, "x", P_::kPadK);
```
[[As future consideration]]: To eliminate the separators 'x' and '_' between strings, we can use function composition with the separators as arguments, as follows: `F(F(Xi, "x"), F(Xi, "x"), "_")` where `F = concat()`
```cpp
concat(x) = concat(concat(Xi, \*sep = *\, 'x'), concat(Yi, 'x'), '_')
```
As a result, the part of the code responsible for passing multiple parameters...
```cpp
concat("gemm_grouped_", prec_str, "_", 
       P_::kMPerBlock, "x", P_::kNPerBlock, "x", P_::kKPerBlock, "_", 
       P_::VectorSizeA, "x", P_::VectorSizeB, "x", P_::VectorSizeC, "_", 
       P_::kPadM, "x", P_::kPadN, "x", P_::kPadK);
```
...could look as follows: either:
```cpp
concat(concat("gemm_grouped_", prec_str),
       concat(P_::kMPerBlock, P_::kNPerBlock, P_::kKPerBlock, "x"),
       concat(P_::VectorSizeA, P_::VectorSizeB, P_::VectorSizeC, "x"),
       concat(P_::kPadM, P_::kPadN, P_::kPadK, "x"), "_")
``` 
or 
```cpp
concat<Sep::_>(concat<Sep::x>("a"), concat<Sep::_>("a", 1));
``` 
where `Sep` is enum..

Going deeper, we could eliminate the multiple explicit uses of concat(). For example, we could... Internal functions usually take the separator "x", so we could avoid multiple external calls, we can use either higher-order function or use recursion, when the params are `initialized_list`
```cpp
concat({{P_::kMPerBlock, P_::kNPerBlock, P_::kKPerBlock},
       {P_::VectorSizeA, P_::VectorSizeB, P_::VectorSizeC},
       {P_::kPadM, P_::kPadN, P_::kPadK}, "x"}, "_")
```

mozga-amd replied:
[[Consider]] To improve readability, it could be looks as follows: (here and in other places)
```cpp
concat("gemm_grouped_", prec_str, "_", 
       P_::kMPerBlock, "x", P_::kNPerBlock, "x", P_::kKPerBlock, "_", 
       P_::VectorSizeA, "x", P_::VectorSizeB, "x", P_::VectorSizeC, "_", 
       P_::kPadM, "x", P_::kPadN, "x", P_::kPadK);
```

mozga-amd commented on include/ck_tile/ops/gemm/kernel/grouped_gemm_kernel.hpp:
```diff
@@ -64,6 +64,25 @@ struct GroupedGemmKernel : public GemmKernel<TilePartitioner_, GemmPipeline_, Ep
 
+    [[nodiscard]] CK_TILE_HOST static const std::string GetName()
+    {
+#define _SS_ std::string
```
[[Consider]] std::string is used only three times in lambda —maybe it's worth considering using it directly and removing the mapping?

mozga-amd commented on include/ck_tile/ops/gemm/kernel/batched_gemm_kernel.hpp:
```diff
@@ -57,6 +58,30 @@ struct BatchedGemmKernel : public GemmKernel<TilePartitioner_, GemmPipeline_, Ep
+            return base_str;
+        }();
+
+        return _SS_("gemm_batched_") + _SS_(prec_str) + "_" +
```
Why can't the `concat()` function be used here?

aledudek replied:
It can be also used here, done

mozga-amd commented on include/ck_tile/ops/gemm/kernel/grouped_gemm_kernel.hpp:
```diff
@@ -64,6 +64,25 @@ struct GroupedGemmKernel : public GemmKernel<TilePartitioner_, GemmPipeline_, Ep
+        // clang-format off
+        using P_ = GemmPipeline;
+
+        auto prec_str = [&] () {
```
[[Consider]]  Could these lambdas be global for all kernels? It looks like that we have to pass 3-data-types

aledudek committed:
[CK_TILE] GetName - small adjustments, format change

aledudek committed:
Merge branch 'develop' into getname_gemm

aledudek committed:
post merge develop fix

aledudek committed:
Remove commented code

illsilin committed:
Merge from internal (#1857)

* enable batched_gemm_softmax_gemm_perm_wmma for gfx12

* disable instances with blocksize=256 in attention examples

* debuggging

* debug

* fixed lds_enabled

* debugging

* Fix and add limit to skiplds feature

* Enable skipLds feature and fix compilation bugs

* add ck_tile definitions for gfx12

* fix clang format and test/wmma_op

* updage instances cmake for gfx12

* disable the test_wmma_op on gfx12

* fix the builds for gfx950

* add gfx12 and gfx950 to default target list

* clean-up cmake file

* Initial introduction of OFP8 data types.

* Renamed FP8 and BF8 tests into FP8_FNUZ and BF8_FNUZ.

* Implementation of ConvertFP32Nearest in test_fp8_ocp.

* Remove dependence on possibly undeclared alias.

* Implement FP8OCP test for stochastic rounding mode.

* Implement FP8OCP tests for half_t type conversions.

* enable bf16 atomic add on gfx950

* Implement ConvertFP32Nearest test.

* Implement ConvertFP32Stochastic test.

* Implement ConvertFP16Nearest and ConvertFP16Stochastic tests.

* Refactoring. Move FP8 definitions into a separate header file.

* Enable easy switching between architectures.

* Fix compilation error for gfx942 architecture.

* Add fp4 type with constants

* only builf gfx950 branch for gfx950 target by default

* Enable OCP build of example_gemm_xdl_fp8.

* Fix formatting.

* fix the build logic for gfx950

* Improve GEMM example verbosity.

* Add constexpr where applicable.

* fix the logic of enabling XDL and WMMA instances

* Improve GEMM example verbosity.

* Enable build of example_gemm_xdl_fp8_bf8 test.

* Fix tests for gfx1101 architecture.

* Build DPP examples only on gfx103 and gfx11 architectures.

* Optionaly run either CPU or GPU verifications with GEMM examples.

* Extend GeneratorTensor_Sequential to produce values of prescribed data types.

* Add missing constructor.

* Add scale type and mxfp conversions

* Update conversions

* Add conversion tests

* Fix typo

* Improve infrastructure for OFP8 data type support.

* BUGFIX. Should not use FP8 as Compute/Accum data type.

* Add custom target for grouped_convnd_bwd_weight tests.

* Can build `tests` target on gfx950.

* Bugfixes on gfx1101 architecture.

* Fix dependencies.

* Add stochastic rounding tests

* Provide single point of truth for FP8 INF and NAN checks

* Prevent instantiation of operators that are not supported by FP8 data types

* Add FP8 type selection into client_axample CMakeLists.txt

* Prevent sccache server from shutting down during build

* Fix test success reporting logic

* Change default verification method to CPU.

GPU verification takes too much time to complete on the emulator.

* Add scale <-> float conversions

* Add scaled conversions with tests

* Add device conversions

* Make sure all tests and examples are built for gfx950

* Facilitate testing of FP8 data types on the emulator

* Introduce two new tensor generators

* Enable instances built for gfx94 to be built on gfx950

* Verify 35_splitk_gemm on floating point numbers.

splitk gemm appears to be losing precision VS reference implementation when FP numbers are involved.

* Format

* Verify 04_gemm_add_add_fastgelu on floating point numbers

* Verify 20_grouped_conv_bwd_weight on floating point numbers

* Verify 38_grouped_conv_bwd_data_multiple_d on floating point numbers

* Verify more tests on floating point data

* Fix data types and improve testing verbocity.

* Add fp4 vectors

* Add debug tests

* Upgrade to NPI 573 build docker.

* Skip on gemm_universal tests.

The tests take too long to complete on the emulator.
Need to see if it is possible to reduce the scope of the testing to just FP8 data types.

* Add new mfma instructions and examples

* Add preprocessor directives for gfx950 specific code

* Fix gfx1101 build

* Document test availability

* Re-enable fp8 gemms for gfx94/95

* Cherry-pick GEMM Universal tests for FP8 data types

* Cleanup

* Add vector types and tests

* Add check_err function

* Add tensor generators

* CK_USE_GFX94 has already been set on this branch

* Fix

* Address formatting issues and leftovers

* Make fail/pass logic consistent within 01_gemm folder

Removed multiple negations in fail/pass logic to propagate `true` as the success indicator.

* Fix GPU verification reporting logic.

* Update year in copyright notice.

* Cleanup

* Use `enum class` instead of `enum`

* Remove set_property for FP8 tests

* Add vector conversions

* Fix

* Fix linker errror

* Clean up

* Fix gfx950 conversions

* Clean up

* Fix more gfx950 conversions

* Fix even more gfx950 conversions

* Narrowing the scope of PR to OCP FP8 enablement only

* Add tests for OCP FP8 vector_type storage

* Fix client examples build

* Fix typo

* Update e8m0 casting

* Rename E8M0 type

* Update unpack method

* Cleanup merge artifacts

* Enable gemm kernel on all gfx9 architectures (#227)

* clean-up

* Implement `non_native_vector_base` with `ext_vector_type` array. (#232)

* Enable support of 1, 2, 4, and 8-byte custom types in CK.

* Fix pool tests for OCP FP8 data type

* Fix build

* Add ckProfiler gemm instances for new mfma instructions and fix ckProfiler build on MI350

* fix clang format

* Add new mfma instructions and examples

* Add preprocessor directives for gfx950 specific code

* Add ckProfiler gemm instances for new mfma instructions and fix ckProfiler build on MI350

* fix clang format

* Fix clang format for the newly merged files

* Use the existing example instances for fp16 bf16 and int8

* Remove comment on new mfma instructions in MfmaInstr

* Update include/ck/tensor_operation/gpu/grid/gridwise_batched_gemm_gemm_xdl_cshuffle_v1.hpp

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* merge from public repo

* Fix ck build

* Fix ck build

* Use double for max_abs_in_val

* Move scaled_type_convert functions to a separate header (#251)

* re-enable building mha lib and gemm_universal_f8 instances for gfx950

* Update library/src/tensor_operation_instance/gpu/CMakeLists.txt

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* fix typo for CK_USE_OCP_FP8

* fix typo for CK_USE_OCP_FP8

* Add FP6 and BF6 types (#261)

* Add a rounding flag

* Add FP6 and BF6

* Add tests

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* Clean up

---------

Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>

* fix one more typo

* Refactor E8M0 scale implementation (#262)

* Refactor E8M0 scale implementation

* Add MXFP6 and MXBF6 conversion methods (#270)

* Add conversions

* Add tests

* Add docstrings

* Add scaled conversions

* Add fp6/bf6 tests

* Remove misleading fp4 test case

* Add docstrings

* Clean up

* Address comments

* Set stricter tolerances for RNE tests

* Add missing tests

* Add native conversions to float

* Revert "Add native conversions to float"

This reverts commit 09467111f73b753c8cc3d597533b187940353dab.

* Update copyright years

* replace the fp6 with bf6 convert calls in test_bf6

* fix test_bf6

* enable smfmac test

* [MX FP8] Add Scaled Type Convert Functions for OCP FP8/BF8 data types (#271)

* Move scaled_type_convert functions to a separate header

* Introduce MX data tests

* Build MX tests only on relevant architectures

* Refactor E8M0 scale implementation

* Fix `config.h` typo

* Cleanup deprecated symbols

* Refactor `amd_ck_fp8.hpp`

* `scaled_type_convert` for `f8_ocp_t`

* Implement test for MX FP8 scaled type convert

* Implement test for MX BF8 scaled type convert

* Scaled type convert for vectors of 2 FP8 elements

* Scaled type convert for vectors of 16 FP8 elements

* Implementation of scaled conversion from F32 to F8

* Add tests for scaled conversions from FP32 to FP8

* Add documentation to the test functions

* Implementation of scaled conversion from F32x2 to F8x2

* Implementation of scaled conversion from F32x16 to F8x16

* Implementation of scaled conversion from F32x32 to F8x32

* Implementation of scaled conversion from F8x32 to F32x32

* Verified on the emulator

* MX FP GEMM - Example Template (#277)

Temporarily uses `DeviceGemmMultiD_ABScale_Xdl_CShuffle_V3` kernel and 128x128 scaling matrices.
Must be modified to use MX-native GEMM kernell with 16 or 32 component vectors per scale.

Verified on the emulator.

* Add vector support

* Add tests

* Add missing type aliases

* Fix test naming

* only build mx example for gfx950

* disable CK_USE_AMD_MFMA_GFX950 by default

* fic build for multiple archs

* fix typo

* fix typo

* Update unpack signature

* Fix merge

* Add size checks in pack function

* Add a flag

* Add conversions

* Fix build logic

* Update pack/unpack methods

* Remove unneeded AsType accessors

* Add docstrings

* Add a flag to config file

* Test the functionality of V_MFMA_F32_16X16X128_F8F6F4 and  V_MFMA_F32_32X32X64_F8F6F4 instructions. (#293)

* Introduced MFMA tests

* Verified f8f6f4 MFMA Instructions

* Move flag logic to scaled_type_convert header

* Use pointers instead of array indices

* Fix a typo

* Update tests and pack functions

* Fix gemm gemm on gfx950

* Fix clang format

* restore the default gput target lists

* fix the jenkinsfile

* add missing ifdef

---------

Co-authored-by: Jing Zhang <jizhan@amd.com>
Co-authored-by: aska-0096 <haocwang@amd.com>
Co-authored-by: Jun Liu <Liu.Jun@amd.com>
Co-authored-by: Andriy Roshchenko <andriy.roshchenko@amd.com>
Co-authored-by: Rostyslav Geyyer <rosty.geyyer@amd.com>
Co-authored-by: Rostyslav Geyyer <46627076+geyyer@users.noreply.github.com>
Co-authored-by: root <root@banff-cyxtera-s83-2.ctr.dcgpu>
Co-authored-by: Andriy Roshchenko <107577548+andriy-ca@users.noreply.github.com>
Co-authored-by: jefyang1 <146495389+jefyang1@users.noreply.github.com>
Co-authored-by: jefyang1 <Jeffreyj.Yang@amd.com>

illsilin committed:
restore cron trigger (#1863)

jakpiase committed:
add vectorloads on non-k dim for memory pipelines (#1856)

kylasa committed:
Support for dtypes (fp8, bf8, bf16 and fp16) for the ck_tile/03_gemm example. (#1845)

* Support bf16/fb8/bf8 datatypes for ck_tile/gemm

* remove commented out code.

* Addressing code review comments and enabling universal_gemm for all the supported data types.

* Merge conflict resolution.

* Solve the memory pipeline compilation error. Merge with the new change of CShuffle

* finish the feature, pass the tests

* Fix the pipeline and add the benchmark script for other data types

---------

Co-authored-by: ThomasNing <thomas.ning@amd.com>

aledudek committed:
Extract prec_str and add separator to concat

aledudek committed:
Merge branch 'develop' into getname_gemm

aledudek committed:
GetName add

aosewski committed:
CK Tile - small fix to hotloop scheduler & KPack value. (#1867)

* Use SmemPack in HotLoop scheduler

* Additional debug print information

* Change KPack value.

Hardcode for now, as without AK1/BK1 there's no good way to determine
its value.

* Fix HotLoopScheduler MFMA instr parameters.

aledudek committed:
Merge branch 'develop' into getname_gemm

aledudek committed:
Merge branch 'develop' into getname_gemm

bartekxk commented on include/ck_tile/ops/gemm/pipeline/gemm_pipeline_agmem_bgmem_creg_v1.hpp:
```diff
@@ -39,6 +40,28 @@ struct GemmPipelineAGmemBGmemCRegV1
+        // clang-format on
+    }
+
+    CK_TILE_HOST_DEVICE static constexpr index_t GetStaticLdsSize()
```
Is it unused?

aledudek replied:
Some leftover from merging with develop probably, will remove it as it's no longer present on develop, thanks!

aledudek committed:
Merge branch 'getname_gemm' of https://github.com/ROCm/composable_kernel into getname_gemm

aledudek committed:
Resolve merge issues

aledudek committed:
Merge branch 'develop' into getname_gemm
